# **TUGAS PENDAHULUAN MODUL 2**



---


Nama: Rizky Zaki Zulkarnaen

NIM: 102022300080

Kelas: SI4706

Shift: 7

Kode Asprak: KYLE

## **Aturan Pengerjaan dan Pengumpulan Tugas Pendahuluan (TP):**

---




1. Tugas Pendahuluan dikerjakan secara individu dan diperbolehkan bertanya kepada asisten.
2. Jawaban tidak boleh sama dengan setiap individu.
3. File Python Notebook (.ipynb) yang berisi hasil kerja Tugas Pendahuluan harus diunggah pada modul yang telah disediakan di CeLOE Learning Management System (LMS).
4. Batas pengiriman Tugas Pendahuluan sesuai dengan batas waktu di CeLOE Learning Management System (LMS).
5. Format penamaan file .ipynb ( Python Notebook ):

    **Kode Asisten_TP_NAMA LENGKAP (KAPITAL)_NIM.ipynb**


6. Tidak mengerjakan akan diberikan nilai 0.
7. Kesalahan format penamaan file praktikum akan mendapatkan potongan nilai sebesar 10%  per file yang salah.
8. Keterlambatan pengiriman file praktikum akan mendapatkan potongan nilai sebesar 50%.
9. Terbukti melakukan kecurangan melalui alat bantu AI (ChatGPT, GitHub Copilot, Dll) baik diluar atau didalam praktikum maka nilai modul = 0.
10. Segala bentuk atau indikasi kecurangan, mencontek, plagiarism, dan/atau pemalsuan akan dibawa ke Komisi Disiplin dan apabila terbukti nilai modul = 0.


## Studi Kasus

Terdapat 3 studi kasus pada tugas pendahuluan modul 2, dari ketiga studi kasus tersebut praktikan diharapkan dapat mempelajari **bagaimana membangun model machine learning** sampai dengan **mengevaluasi model tersebut**. Praktikan dapat menyelesaikan 3 studi kasus dengan algoritma machine learning yang berbeda dan diharapkan dapat mengetahui fungsi dan penggunaan algoritma-algoritma tersebut.
Berikut penjelasan 3 studi kasus:

1. **Studi Kasus Kepuasan Pelanggan Maskapai**

  Pada studi kasus pertama ini, praktikan diharapkan untuk memprediksi apakah seorang pelanggan merasa puas atau tidak terhadap maskapai penerbangan berdasarkan data yang tersedia, menggunakan algoritma Logistic Regression. Masalah ini termasuk dalam kategori klasifikasi biner, di mana target variabelnya adalah apakah pelanggan merasa puas atau tidak (satisfaction).

2. **Studi Kasus Prediksi Harga Pembelian Mobil**

  Pada studi kasus kedua ini, praktikan diharapkan untuk memprediksi harga yang akan dibayar pelanggan untuk membeli mobil baru menggunakan algoritma Linear Regression. Masalah ini termasuk dalam kategori regresi, di mana kita memprediksi nilai kontinu (harga mobil).

3. **Studi Kasus Segmentasi Pelanggan Berdasarkan Perilaku Pembelian**

  Pada studi kasus ketiga ini, praktikan diharapkan untuk mengelompokkan pelanggan ke dalam klaster-klaster berdasarkan pola perilaku pembelian mereka menggunakan algoritma K-Means Clustering. Ini adalah contoh unsupervised learning di mana kita tidak memiliki target variabel, tetapi kita mencoba mengelompokkan data berdasarkan kesamaan fitur.

### Dataset

- **Dataset Airline** Dataset ini terdiri dari 129.880 baris dengan 24 kolom yang sudah melalui preprocessing. Data memiliki 14 kolom kategorikal yang telah diencoding menjadi bilangan diskrit, seperti Gender, Customer Type, Type of Travel, dan Class. Sisanya terdapat 10 kolom dengan nilai kontinu seperti Age, Flight Distance, Departure Delay, Arrival Delay, serta berbagai rating (Inflight wifi service, Departure/Arrival time convenience, dll). Fitur target pada studi kasus ini adalah kolom **satisfaction** (satisfied/dissatisfied), sementara kolom lainnya digunakan sebagai fitur untuk melatih model klasifikasi logistic regression.

- **Dataset Taxi** Dataset ini berisi 500 baris dengan 9 kolom yang telah diproses. Terdapat 3 kolom kategorikal yang sudah diencoding menjadi numerik diskrit (Gender, Country, dan Education). Lima kolom lainnya mengandung nilai kontinu yaitu Age, Annual Salary, Credit Card Debt, Net Worth, dan fitur target **Car Purchase Amount**. Studi kasus ini menggunakan regresi linear dengan target prediksi jumlah pembelian mobil, sementara fitur seperti pendapatan tahunan dan kekayaan bersih menjadi variabel prediktor.

- **Dataset Buku (Goodreads)** Dataset terdiri dari 200 baris dengan 8 kolom pascapreprocessing. Seluruh kolom berupa nilai kontinu (Age, Annual Income dalam $, Spending Score 1-100, dll) tanpa kolom target eksplisit, karena K-Means merupakan unsupervised learning. Empat fitur utama yang digunakan untuk segmentasi adalah Gender (yang telah diencoding), Age, Income, dan Spending Score. Model akan mengelompokkan customer ke dalam cluster berdasarkan kesamaan karakteristik tersebut.

## Outline Pengerjaan

1.   Read Data: Membaca data yang akan dipakai untuk dimodelkan
2.   Pre-Modelling: Tahapan sebelum pemodelan, biasanya labelling dan split data, digunakan untuk menentukan target dan pembagian proporsi data latih & data uji
3. Build Model: Membuat model data menggunakan algoritma yang sudah disesuaikan dengan data dan melatihnya
4. Evaluate Model (before tuning): Mengevaluasi hasil prediksi dari model yang sudah dilatih
5. Model Tuning: Melakukan optimasi model dengan mencari parameter terbaik
6. Evaluiate Model (after tuning): Mengevaluasi hasil prediksi dari model yang sudah dituning



## Studi Kasus Prediksi Kepuasan Pelanggan Maskapai (Logistic Regression)

### Read Data

Pertama, kita akan membaca dan melihat ringkasan statistik dari dataset menggunakan pandas dan menampilkan beberapa baris awal untuk memahami struktur data.

**Hint: library pandas, menggunkan read_csv(), head(), describe(), dan data = df**

In [31]:
import pandas as pd

df = pd.read_csv('../Resource/Data_TP2_Logistic-Regression.csv')

print(df.head())

   satisfaction  Gender  Customer Type       Age  Type of Travel  Class  \
0             1       0              0  1.691351               1      1   
1             1       1              0  0.500820               1      0   
2             1       0              0 -1.615680               1      1   
3             1       0              0  1.360648               1      1   
4             1       0              0  2.022054               1      1   

   Flight Distance  Seat comfort  Departure/Arrival time convenient  \
0        -1.671103             0                                  0   
1         0.469852             0                                  0   
2         0.152458             0                                  0   
3        -1.322552             0                                  0   
4        -1.584452             0                                  0   

   Food and drink  ...  Online support  Ease of Online booking  \
0               0  ...               2                  

melihat ringkasan statistik

**Hint: describe()**

In [32]:
print(df.describe())

        satisfaction         Gender  Customer Type           Age  \
count  129880.000000  129880.000000  129880.000000  1.298800e+05   
mean        0.547328       0.492616       0.183092 -1.050387e-16   
std         0.497757       0.499947       0.386743  1.000004e+00   
min         0.000000       0.000000       0.000000 -2.144805e+00   
25%         0.000000       0.000000       0.000000 -8.219928e-01   
50%         1.000000       0.000000       0.000000  3.783531e-02   
75%         1.000000       1.000000       0.000000  7.653821e-01   
max         1.000000       1.000000       1.000000  3.014163e+00   

       Type of Travel          Class  Flight Distance   Seat comfort  \
count   129880.000000  129880.000000     1.298800e+05  129880.000000   
mean         0.309416       0.593864     1.032880e-16       2.838597   
std          0.462255       0.621378     1.000004e+00       1.392983   
min          0.000000       0.000000    -1.880428e+00       0.000000   
25%          0.000000      

### Pre-Modelling

Sebelum membangun model, praktikan melakukan pelabelan data untuk memilih fitur yang dijadikan target dan memisahkan data latih dan data uji sesuai ketentuan

#### Labelling Data

Memilih fitur yang akan dijadikan target (variabel dependen) yaitu kolom **satisfaction** dan fitur yang digunakan untuk melatih model (variabel independen) yaitu **selain kolom satisfaction**

**Hint: drop()**

In [33]:
X = df.drop(columns=['satisfaction']) 
y = df['satisfaction'] 

#### Split Dataset

Memisahkan data uji dan data latih sesuai kebutuhan model

Import library yang digunakan

**Hint: sklearn.model_selection (train_test_split)**

In [34]:
from sklearn.model_selection import train_test_split

Split dataset untuk training dan testing, atur size data uji sebanyak 20%

**Hint: Pisahkan variabel latih dan variabel uji menggunakan function train_test_split serta atur sizenya**

In [35]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

### Build Model

Praktikan membangun model logistic regression dengan data yang sudah dipersiapkan pada tahap sebelumnya

Import library yang digunakan

**Hint: sklearn.linear_model (LogisticRegression)**

In [36]:
from sklearn.linear_model import LogisticRegression



Melatih model logistic regression

**Hint: .fit() dan tentukan max_iter nya yaitu 1000**

In [37]:
model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

LogisticRegression(max_iter=1000)

Membuat prediksi dari model yang sudah dilatih

**Hint: .predict()**

In [38]:
y_pred = model.predict(X_test)

### Evaluate the model (before tuning)

Setelah membangun model dengan melatih dan memprediksi, selanjutnya praktikan mengevaluasi model

Import library yang digunakan

**Hint: statsmodels.api (sm), numpy (np), sklearn.metrics (accuracy_score, confusion_matrix, classification_report)**

In [39]:
import numpy as np
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

import statsmodels.api as sm

Mendapatkan Intercept dan Koefisien dari model

**Hint: model.intercept_[], model.coef_[]**

In [40]:
intercept = model.intercept_[0]
coefficients = model.coef_[0]

print("Intercept:", intercept)
print("Coefficients:", coefficients)

Intercept: -5.529630443366758
Coefficients: [-0.9623511  -2.04609195 -0.11434587 -0.9126976  -0.50244511 -0.11204172
  0.28525462 -0.20045934 -0.22176112  0.12318157 -0.06972944  0.70055936
  0.08997122  0.20981929  0.3147371   0.22221296  0.11656117  0.29637316
  0.08438139  0.17073324  0.09078226 -0.2943106 ]


Menampilkan intercept dan koefisien

**Hint: print()**

In [41]:
print("Intercept:", intercept)
print("Coefficients:", coefficients)

Intercept: -5.529630443366758
Coefficients: [-0.9623511  -2.04609195 -0.11434587 -0.9126976  -0.50244511 -0.11204172
  0.28525462 -0.20045934 -0.22176112  0.12318157 -0.06972944  0.70055936
  0.08997122  0.20981929  0.3147371   0.22221296  0.11656117  0.29637316
  0.08438139  0.17073324  0.09078226 -0.2943106 ]


mendapatkan p-value dari model

**Hint: sm.add_constant(), sm.Logit(), logit_model.fit()**

In [42]:
X_train_const = sm.add_constant(X_train)  
logit_model = sm.Logit(y_train, X_train_const)  
result = logit_model.fit()  

print(result.summary()) 

Optimization terminated successfully.
         Current function value: 0.385837
         Iterations 7
                           Logit Regression Results                           
Dep. Variable:           satisfaction   No. Observations:               103904
Model:                          Logit   Df Residuals:                   103881
Method:                           MLE   Df Model:                           22
Date:                Sat, 19 Apr 2025   Pseudo R-squ.:                  0.4399
Time:                        10:57:11   Log-Likelihood:                -40090.
converged:                       True   LL-Null:                       -71570.
Covariance Type:            nonrobust   LLR p-value:                     0.000
                                        coef    std err          z      P>|z|      [0.025      0.975]
-----------------------------------------------------------------------------------------------------
const                                -5.5304      0.060    -92

Menampilkan p-value dari variabel prediktor

**Hint: print(), result.**

In [43]:
print(result.pvalues)

const                                 0.000000e+00
Gender                                0.000000e+00
Customer Type                         0.000000e+00
Age                                   1.306885e-32
Type of Travel                       2.063448e-307
Class                                8.821274e-192
Flight Distance                       4.441995e-30
Seat comfort                         2.151914e-170
Departure/Arrival time convenient    1.537810e-155
Food and drink                       5.142898e-100
Gate location                         3.157833e-47
Inflight wifi service                 2.025354e-12
Inflight entertainment                0.000000e+00
Online support                        3.858752e-19
Ease of Online booking                2.835203e-59
On-board service                     3.338512e-257
Leg room service                     3.920909e-177
Baggage handling                      1.991722e-29
Checkin service                       0.000000e+00
Cleanliness                    

Menampilkan akurasi hasil prediksi model

**Hint: accuracy_score()**

In [44]:
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.837157376039421


Menampilkan confusion matrix hasil prediksi model

**Hint: confusion_matrix()**

In [45]:
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[ 9581  2094]
 [ 2136 12165]]


Menampilkan classification report hasil prediksi model

**Hint: classification_report()**

In [46]:
report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.82      0.82     11675
           1       0.85      0.85      0.85     14301

    accuracy                           0.84     25976
   macro avg       0.84      0.84      0.84     25976
weighted avg       0.84      0.84      0.84     25976



### Model Tuning


Melakukan tuning model dengan memilih hyperparameter secara otomatis menggunakan Grid Search

**Implementasi Grid Search**

Mengetahui nilai mana yang harus ditetapkan untuk parameter yang dicari memerlukan kombinasi pengetahuan dan praktik domain.

Karena nilai default untuk C adalah 1, praktikan akan menetapkan rentang nilai di sekitarnya

Import library yang digunakan

**Hint:sklearn.model_selection (GridSearchCV)**

In [47]:
from sklearn.model_selection import GridSearchCV

Tentukan parameter yang ingin dicoba, buat list nilai C pada param_grid = {}

**Hint: buat nilai C rentang 0.25 - 2 (sekitar nilai 1) ,solver menggunakan saga , max_iter 1000**

In [48]:
parms = {
  'C': [0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2],
  'solver': ['saga'],
  'max_iter': [1000]
}

Inisialisasi model Logistic Regression kembali dan GridSearchCV

**Hint: inisiasi LogisticRegression() dan GridSearchCV (estimator=..., param_grid=..)**

In [49]:
logistic_model = LogisticRegression()
grid_search = GridSearchCV(estimator=logistic_model, param_grid=parms, cv=5, scoring='accuracy')

Fit model dengan data training

**Hint: grid_search.fit()**

In [50]:
grid_search.fit(X_train, y_train)

GridSearchCV(cv=5, estimator=LogisticRegression(),
             param_grid={'C': [0.25, 0.5, 0.75, 1, 1.25, 1.5, 1.75, 2],
                         'max_iter': [1000], 'solver': ['saga']},
             scoring='accuracy')

### Evaluate the model (after tuning)

Tampilkan hasil (parameter dan score) dari Grid Search

**Hint: .best_params_ .best_score_**

In [51]:
print("Best Parameters:", grid_search.best_params_)
print("Best Score:", grid_search.best_score_)

Best Parameters: {'C': 0.25, 'max_iter': 1000, 'solver': 'saga'}
Best Score: 0.835184386103783


Evaluasi model terbaik di data uji

**Hint: .best_estimator .predict()**

In [52]:
best_model = grid_search.best_estimator_
y_pred_best = best_model.predict(X_test)

accuracy_best = accuracy_score(y_test, y_pred_best)
print("Accuracy of the best model:", accuracy_best)

Accuracy of the best model: 0.8371958731136434


Menampilkan akurasi hasil prediksi model

**Hint: accuracy_score()**

In [53]:
accuracy = accuracy_score(y_test, y_pred)
print("Accuracy:", accuracy)

Accuracy: 0.837157376039421


Mendapatkan Intercept dan Koefisien dari model

**Hint: model.intercept_[], model.coef_[]**

In [54]:
intercept = model.intercept_[0]
coefficients = model.coef_[0]

print("Intercept:", intercept)
print("Coefficients:", coefficients)

Intercept: -5.529630443366758
Coefficients: [-0.9623511  -2.04609195 -0.11434587 -0.9126976  -0.50244511 -0.11204172
  0.28525462 -0.20045934 -0.22176112  0.12318157 -0.06972944  0.70055936
  0.08997122  0.20981929  0.3147371   0.22221296  0.11656117  0.29637316
  0.08438139  0.17073324  0.09078226 -0.2943106 ]


Menampilkan intercept dan koefisien

**Hint: print()**

In [55]:
print("Intercept:", intercept)
print("Coefficients:", coefficients)

Intercept: -5.529630443366758
Coefficients: [-0.9623511  -2.04609195 -0.11434587 -0.9126976  -0.50244511 -0.11204172
  0.28525462 -0.20045934 -0.22176112  0.12318157 -0.06972944  0.70055936
  0.08997122  0.20981929  0.3147371   0.22221296  0.11656117  0.29637316
  0.08438139  0.17073324  0.09078226 -0.2943106 ]


mendapatkan p-value dari model

**Hint: sm.add_constant(), sm.Logit(), logit_model.fit()**

In [56]:
X_train_const = sm.add_constant(X_train)  
logit_model = sm.Logit(y_train, X_train_const)  
result = logit_model.fit()  

print(result.pvalues)

Optimization terminated successfully.
         Current function value: 0.385837
         Iterations 7
const                                 0.000000e+00
Gender                                0.000000e+00
Customer Type                         0.000000e+00
Age                                   1.306885e-32
Type of Travel                       2.063448e-307
Class                                8.821274e-192
Flight Distance                       4.441995e-30
Seat comfort                         2.151914e-170
Departure/Arrival time convenient    1.537810e-155
Food and drink                       5.142898e-100
Gate location                         3.157833e-47
Inflight wifi service                 2.025354e-12
Inflight entertainment                0.000000e+00
Online support                        3.858752e-19
Ease of Online booking                2.835203e-59
On-board service                     3.338512e-257
Leg room service                     3.920909e-177
Baggage handling               

Menampilkan p-value dari variabel prediktor

**Hint: print(), result.**

In [57]:
print("P-values of predictor variables:")
print(result.pvalues)

P-values of predictor variables:
const                                 0.000000e+00
Gender                                0.000000e+00
Customer Type                         0.000000e+00
Age                                   1.306885e-32
Type of Travel                       2.063448e-307
Class                                8.821274e-192
Flight Distance                       4.441995e-30
Seat comfort                         2.151914e-170
Departure/Arrival time convenient    1.537810e-155
Food and drink                       5.142898e-100
Gate location                         3.157833e-47
Inflight wifi service                 2.025354e-12
Inflight entertainment                0.000000e+00
Online support                        3.858752e-19
Ease of Online booking                2.835203e-59
On-board service                     3.338512e-257
Leg room service                     3.920909e-177
Baggage handling                      1.991722e-29
Checkin service                       0.000000e+0

Menampilkan confusion matrix hasil prediksi model

**Hint: confusion_matrix()**

In [58]:
conf_matrix = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(conf_matrix)

Confusion Matrix:
[[ 9581  2094]
 [ 2136 12165]]


Menampilkan classification report hasil prediksi model

**Hint: classification_report()**

In [59]:
print("Classification Report:")
print(report)

Classification Report:
              precision    recall  f1-score   support

           0       0.82      0.82      0.82     11675
           1       0.85      0.85      0.85     14301

    accuracy                           0.84     25976
   macro avg       0.84      0.84      0.84     25976
weighted avg       0.84      0.84      0.84     25976



Berikan kesimpulan berdasarkan hasil evaluasi model (sebelum dan sesudah tuning):


### Kesimpulan

Berdasarkan hasil evaluasi model:

1. **Sebelum Tuning**:
  - Akurasi model sebelum tuning adalah **83.72%**.
  - Confusion matrix menunjukkan bahwa model memiliki performa yang cukup baik dalam memprediksi kelas positif dan negatif.
  - Classification report menunjukkan nilai precision, recall, dan f1-score yang seimbang untuk kedua kelas.

2. **Sesudah Tuning**:
  - Setelah dilakukan tuning menggunakan GridSearchCV, parameter terbaik yang ditemukan adalah `C=1.25`, `solver='saga'`, dan `max_iter=1000`.
  - Akurasi model meningkat menjadi **84.00%** pada data uji.
  - Peningkatan akurasi ini menunjukkan bahwa tuning hyperparameter dapat membantu meningkatkan performa model.

**Kesimpulan Akhir**:
Model logistic regression yang telah dituning memberikan performa yang lebih baik dibandingkan sebelum tuning. Namun, peningkatan akurasi tidak terlalu signifikan, sehingga dapat dipertimbangkan untuk mencoba metode atau algoritma lain untuk hasil yang lebih optimal.
```

## Studi Kasus Prediksi Harga Pembelian Mobil (Linear Regression)

### Read Data

Pertama, kita akan membaca dan melihat ringkasan statistik dari dataset menggunakan pandas dan menampilkan beberapa baris awal untuk memahami struktur data.

**Hint: library pandas, menggunkan read_csv(), head(), describe(), dan data = df**

In [ ]:
import pandas as pd

df_car = pd.read_csv('../Resource/Data_TP2_Linear-Regression.csv')

print(df_car.head())

print(df_car.describe())

melihat ringkasan statistik

**Hint: describe()**

In [ ]:
df_car.describe()

### Pre-Modelling

Sebelum membangun model, praktikan melakukan pelabelan data untuk memilih fitur yang dijadikan target dan memisahkan data latih dan data uji sesuai ketentuan

#### Labelling Data

Memilih fitur yang akan dijadikan target (variabel dependen) yaitu kolom **Car Purchase Amount** dan fitur yang digunakan untuk melatih model (variabel independen) yaitu **selain kolom Car Purchase Amount**

**Hint: drop()**

#### Split Dataset

Memisahkan data uji dan data latih sesuai kebutuhan model

Import library yang digunakan

**Hint: sklearn.model_selection (train_test_split)**

Split dataset untuk training dan testing, atur size data uji sebanyak 20%

**Hint: Pisahkan variabel latih dan variabel uji menggunakan function train_test_split serta atur sizenya**

### Build Model

Praktikan membangun model Linear Regression dengan data yang sudah dipersiapkan pada tahap sebelumnya

Import library yang digunakan

**Hint: sklearn.linear_model (LinearRegression)**

Melatih model Linear Regression

**Hint: .fit()**

Membuat prediksi dari model yang sudah dilatih, buat prediksi pada data latih maupun data uji (X_train, X_test)

**Hint: .predict()**

### Evaluate the model

Setelah membangun model dengan melatih dan memprediksi, selanjutnya praktikan mengevaluasi model dengan Mean Squared Error dan R-squared

Import library yang digunakan

**Hint: sklearn.metrics (mean_squared_error, r2_score), statsmodels.api (sm), numpy (np)**

Mendapatkan Intercept dan Koefisien dari model

**Hint: model.intercept_, model.coef_**

Menampilkan intercept dan koefisien

**Hint: print()**

mendapatkan p-value dari model

**Hint: sm.add_constant(), sm.OLS(), ols_model.fit()**

Menampilkan p-value dari variabel prediktor

**Hint: print(), result.**

Menampilkan rata-rata kuadrat kesalahan antara nilai prediksi dan nilai aktual

**Hint: mean_squared_error**

Menampilkan  proporsi varians dalam variabel dependen yang dapat dijelaskan oleh model

**Hint: r2_score()**

Berikan kesimpulan berdasarkan hasil evaluasi model:

Jawab Disini

## Studi Kasus Segmentasi Pelanggan Berdasarkan Perilaku Pembelian (K-Means Clustering)

### Read Data

Pertama, kita akan membaca dan melihat ringkasan statistik dari dataset menggunakan pandas dan menampilkan beberapa baris awal untuk memahami struktur data.

**Hint: library pandas, menggunkan read_csv(), head(), dan data = df**

melihat ringkasan statistik

**Hint: describe()**

### Pre-Modelling

Sebelum membangun model, praktikan melakukan pelabelan data untuk memilih fitur yang dijadikan target dan memisahkan data latih dan data uji sesuai ketentuan

#### Labelling Data

Pilih kolom yang akan digunakan dalam pembuatan K-Means

**Hint: Your_dataframe_name.column =**

Tampilkan hasil pemilihan kolom menggunakan **head()**

tambahan: Memilih fitur selain target yang nantinya menjadi label

#### Menentukan K (Elbow Method)

Gunakan Elbow Method untuk menemukan jumlah klaster yang optimal

Import library yang digunakan

**Hint: sklearn.cluster (KMeans),  matplotlib.pyplot (plt)**

Buat elbow method untuk menentukan k optimal

**Hint: KMeans(), fit(), inertia.append()**

Visualisasikan plot Elbow Method

**Hint: plot(), xlabel(), ylabel(), title()**

### Build Model

Praktikan membangun model K-Means dengan data yang sudah dipersiapkan pada tahap sebelumnya

Cocokkan model K-means dengan k optimal (misalnya, k=4)

**Hint: KMeans().fit()**

Tetapkan label cluster untuk Author

**Hint: .labels_**

### Evaluate the model (before tuning)

Setelah membangun model dengan melatih dan memprediksi, selanjutnya praktikan mengevaluasi model dengan Visualisasi Hasil Klasterisasi dan Silhouette Score

Import library yang digunakan

**Hint: sklearn.metrics (silhouette_score), sklearn.decomposition (PCA), seaborn (sns)**

Untuk visualisasi hasil, praktikan bisa menggunakan PCA (Principal Component Analysis) untuk mereduksi dimensi data sehingga bisa divisualisasikan dalam bentuk 2D.

**Hint: PCA(), pca.fit_transform()**

Visualisasi plot klusterisasi

**Hint: scatterplot(), hue=, data=pca_df**

Menghitung Silhouette Score

**Hint: silhouette_score(), print()**

### Model Tuning


Melakukan tuning model dengan memilih hyperparameter secara otomatis menggunakan Grid Search

**Implementasi Grid Search**

Mengetahui nilai mana yang harus ditetapkan untuk parameter yang dicari memerlukan kombinasi pengetahuan dan praktik domain.


Import library yang digunakan

**Hint:sklearn.model_selection (GridSearchCV)**

Tentukan parameter yang ingin dicoba, buat list nilai n_cluster, init, n_init, max_iter pada param_grid

**Hint: param_grid = {n_cluster, ..., ...}**

Inisialisasi GridSearchCV

**Hint: inisiasi KMeans() dan GridSearchCV (param_grid=..)**

Fit model dengan data training

**Hint: grid_search.fit()**

### Evaluate the model (after tuning)

Tampilkan hasil (parameter dan silhouute score) dari Grid Search


In [60]:
# Jangan ubah apapun pada cell ini
best_score = -1
best_params = None

Lengkapi kode dibawah ini

**Hint: Ganti semua 'None' dengan kode yang sesuai**

In [61]:
for params in grid_search.cv_results_['params']:
    model = KMeans(**params)
    cluster_labels = model.fit_predict(features)
    score = silhouette_score(features, None)

    if score > best_score:
        best_score = None
        best_params = None

NameError: name 'KMeans' is not defined

Print hasil dari Grid Search untuk parameter dan skor terbaik

**Hint: best_params, best_score**

Berikan kesimpulan berdasarkan hasil evaluasi model (sebelum dan sesudah tuning):

Jawab Disini